evaluator/autoware_trajectory_evaluator/src/trajectory_evaluator_node.cpp

```diff
@@ -0,0 +1,234 @@
+{
+  auto current_time = now();
+
+  const auto ego_state_ptr = kinematic_state_sub_.takeData();
+  const auto traj_msg = traj_sub_.takeData();
+  store_trajectory(traj_msg, ego_state_ptr);
+  on_kinematic_state(ego_state_ptr);
+
+  metrics_msg_.stamp = now();
+  metrics_pub_->publish(metrics_msg_);
+  metrics_msg_ = MetricArrayMsg{};
+}
+
+void TrajectoryEvaluatorNode::store_trajectory(
+  const autoware_planning_msgs::msg::Trajectory::ConstSharedPtr traj_msg,
+  const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg)
+{
+  if (traj_msg != nullptr) {
+    TrajectoryWithTimestamp trajectory_with_timestamp;
+    trajectory_with_timestamp.time_stamp = this->now();
+    const auto current_position = odom_msg->pose.pose.position;
+    std::vector<TrajectoryPointWithTime> trajectory_with_time;
+    for (const auto & point : traj_msg->points) {
+      TrajectoryPointWithTime point_with_time(
+        point, rclcpp::Duration::from_seconds(0), trajectory_with_timestamp.time_stamp);
+      trajectory_with_timestamp.trajectory_points.push_back(point_with_time);
+    }
+
+    calculate_time_from_start(trajectory_with_timestamp, current_position, 1e-3);
+    trajectory_history_.push_back(trajectory_with_timestamp);
+    if (trajectory_history_.size() > traj_history_limit) {
+      trajectory_history_.erase(trajectory_history_.begin());
+    }
+  }
+}
+
+void TrajectoryEvaluatorNode::calculate_time_from_start(
+  TrajectoryWithTimestamp & trajectory, const geometry_msgs::msg::Point & current_ego_point,
+  const float min_velocity)
+{
+  auto & trajectory_points = trajectory.trajectory_points;
+  std::vector<autoware_planning_msgs::msg::TrajectoryPoint> trajectory_points_sequence;
+  for (const auto & traj_point_with_time : trajectory_points) {
+    trajectory_points_sequence.push_back(traj_point_with_time.trajectory_point);
+  }
+
+  const auto nearest_segment_idx =
+    autoware::motion_utils::findNearestSegmentIndex(trajectory_points_sequence, current_ego_point);
+
+  if (nearest_segment_idx + 1 >= trajectory_points.size()) {
+    return;
+  }
+
+  for (auto & point_with_time : trajectory_points) {
+    point_with_time.time_from_start = rclcpp::Duration::from_seconds(0);
+  }
+
+  for (size_t idx = nearest_segment_idx + 1; idx < trajectory_points.size(); ++idx) {
+    const auto & from = trajectory_points[idx - 1];
+    const auto velocity = std::max(min_velocity, from.trajectory_point.longitudinal_velocity_mps);
+
+    if (velocity != 0.0) {
+      auto & to = trajectory_points[idx];
+      const double distance = autoware::universe_utils::calcDistance2d(
+        from.trajectory_point.pose, to.trajectory_point.pose);
+      const double time_delta = distance / velocity;
+      to.time_from_start =
+        rclcpp::Duration::from_seconds(time_delta) + trajectory_points[idx - 1].time_from_start;
+    }
+  }
+}
+
+void TrajectoryEvaluatorNode::on_kinematic_state(
+  const nav_msgs::msg::Odometry::ConstSharedPtr odom_msg)
+{
+  if (odom_msg != nullptr) {
```

Comment:
Instead of having the whole function in the `if` block, you should do an early return.
```suggestion
  if (odom_msg == nullptr) {
    return;
  }
```